kale/embed/gripnet.py

```diff
@@ -241,3 +245,312 @@ def forward(
+        """
+
+        if self.setting.concat_output:
+            tmp = []
+            tmp.append(x)
+
+        if self.if_multirelational:
+            if edge_type is None or range_list is None:
+                error_msg = "Both edge_type and range_list are required for multi-relational supervertex."
+                logging.error(error_msg)
+                raise ValueError(error_msg)
+
+        # internal feature aggregation layers
+        for net in self.inter_agg_layers[:-1]:
+            x = (
+                net(x, edge_index, edge_type, range_list)
+                if self.if_multirelational
+                else net(x, edge_index, edge_weight)
+            )
+            x = F.relu(x, inplace=True)
+            if self.setting.concat_output:
+                tmp.append(x)
+
+        x = (
+            self.inter_agg_layers[-1](x, edge_index, edge_type, range_list)
+            if self.if_multirelational
+            else self.inter_agg_layers[-1](x, edge_index, edge_weight)
+        )
+
+        x = F.relu(x, inplace=True)
+        if self.setting.concat_output:
+            tmp.append(x)
+            x = torch.cat(tmp, dim=1)
+
+        return x
+
+    def __repr__(self):
+        tmp = [f"\n    ({i}): {l}" for i, l in enumerate(self.inter_agg_layers)]
+
+        return "{}: ModuleList(\n  (0): InternalFeatureLayer: Embedding({}, {})\n  (1): InternalFeatureAggerationModule: ModuleList({}\n  )\n)".format(
+            self.__class__.__name__, self.in_channels, self.setting.internal_feat_channels, "".join(tmp)
+        )
+
+
+class GripNetExternalModule(Module):
+    """The internal module of a supervertex, which is an external feature layer.
+
+    Args:
+        in_channels (int): Size of each input sample. In GripNet, it shold be the dimension of the output embedding of the
+            corresponding parent supervertex.
+        out_channels (int): Size of each output sample. In GripNet, it is the dimension of the output embedding of
+            the supervertex.
+        num_out_node (int): the number of output nodes.
+    """
+
+    def __init__(self, in_channels: int, out_channels: int, num_out_node: int) -> None:
+        super(GripNetExternalModule, self).__init__()
+
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+        self.num_out_node = num_out_node
+
+        self.exter_agg_layer = GCNEncoderLayer(in_channels, out_channels, cached=True)
+
+    def forward(self, x: torch.Tensor, edge_index: torch.Tensor, edge_weight: torch.Tensor = None, relu=True):
+        """
+        Args:
+            x (torch.Tensor): the input node feature embedding.
+            edge_index (torch.Tensor): edge index in COO format with shape [2, #edges].
+            edge_weight (torch.Tensor, optional): one-dimensional weight for each edge. Defaults to None.
+            relu (bool, optional): if use ReLU before returning node feature embeddings. Defaults to True.
```

Comment:
'use_relu' may be a better variable name for bool 
if use --> whether to use
check ALL bool variable descriptions for 'if'